UnicodeDecodeError on failing assert if non-ASCII characters in __repr__

If a UTF-8 encoded string is returned in the `__repr__` method of a class, a `UnicodeDecodeError` is raised if an `assert` fails while checking any `unicode` attribute on an object of that class (even ASCII-decodable ones). Non-`unicode` attributes (including byte strings) work as expected.
```python
# -*- coding: utf-8 -*-

class A(object):
    name = u"xyzä"
    unicode_attr = u"a"
    ascii_str_attr = "b"
    utf8_str_attr = u"ä".encode("utf-8")

    def __repr__(self):
        return self.name.encode("utf-8")

def test_unexpected_unicodedecodeerror():
    a = A()
    assert not a.name

def test_unexpected_unicodedecodeerror2():
    a = A()
    assert not a.unicode_attr

def test_expected_assertionerror():
    a = A()
    assert not a.ascii_str_attr

def test_expected_assertionerror2():
    a = A()
    assert not a.utf8_str_attr
```
```python
============================= test session starts ==============================
platform darwin -- Python 2.7.15, pytest-3.8.0, py-1.6.0, pluggy-0.7.1
rootdir: /Users/dtomas/work, inifile:
collected 4 items                                                              

test_unicode.py FFFF                                                     [100%]

=================================== FAILURES ===================================
______________________ test_unexpected_unicodedecodeerror ______________________

    def test_unexpected_unicodedecodeerror():
        a = A()
>       assert not a.name
E       UnicodeDecodeError: 'ascii' codec can't decode byte 0xc3 in position 3: ordinal not in range(128)

test_unicode.py:14: UnicodeDecodeError
_____________________ test_unexpected_unicodedecodeerror2 ______________________

    def test_unexpected_unicodedecodeerror2():
        a = A()
>       assert not a.unicode_attr
E       UnicodeDecodeError: 'ascii' codec can't decode byte 0xc3 in position 3: ordinal not in range(128)

test_unicode.py:18: UnicodeDecodeError
_________________________ test_expected_assertionerror _________________________

    def test_expected_assertionerror():
        a = A()
>       assert not a.ascii_str_attr
E       AssertionError: assert not 'b'
E        +  where 'b' = xyzä.ascii_str_attr

test_unicode.py:22: AssertionError
________________________ test_expected_assertionerror2 _________________________

    def test_expected_assertionerror2():
        a = A()
>       assert not a.utf8_str_attr
E       AssertionError: assert not '\xc3\xa4'
E        +  where '\xc3\xa4' = xyzä.utf8_str_attr

test_unicode.py:26: AssertionError
=========================== 4 failed in 0.04 seconds ===========================
```
```
$ pip list
Package        Version
-------------- -------
atomicwrites   1.2.1  
attrs          18.2.0 
funcsigs       1.0.2  
more-itertools 4.3.0  
pathlib2       2.3.2  
pip            18.0   
pluggy         0.7.1  
py             1.6.0  
pytest         3.8.0  
scandir        1.9.0  
setuptools     40.4.1 
six            1.11.0 
wheel          0.31.1 
```
Might be related to https://github.com/pytest-dev/pytest/issues/877 and https://github.com/pytest-dev/pytest/issues/678.

--- a/changelog/3999.bugfix.rst
+++ b/changelog/3999.bugfix.rst
@@ -1,0 +1,1 @@
+Fix ``UnicodeDecodeError`` in python2.x when a class returns a non-ascii binary ``__repr__`` in an assertion which also contains non-ascii text.

--- a/src/_pytest/assertion/rewrite.py
+++ b/src/_pytest/assertion/rewrite.py
@@ -8,6 +8,7 @@ import marshal
 import os
 import re
 import six
+import string
 import struct
 import sys
 import types
@@ -466,10 +467,14 @@ def _saferepr(obj):
 
     """
     r = py.io.saferepr(obj)
-    if isinstance(r, six.text_type):
-        return r.replace(u"\n", u"\\n")
-    else:
-        return r.replace(b"\n", b"\\n")
+    # only occurs in python2.x, repr must return text in python3+
+    if isinstance(r, bytes):
+        # Represent unprintable bytes as `\x##`
+        r = u"".join(
+            u"\\x{:x}".format(ord(c)) if c not in string.printable else c.decode()
+            for c in r
+        )
+    return r.replace(u"\n", u"\\n")
 
 
 from _pytest.assertion.util import format_explanation as _format_explanation  # noqa

--- a/testing/test_assertrewrite.py
+++ b/testing/test_assertrewrite.py
@@ -1,3 +1,4 @@
+# -*- coding: utf-8 -*-
 from __future__ import absolute_import, division, print_function
 
 import glob
@@ -57,7 +58,7 @@ def getmsg(f, extra_ns=None, must_pass=False):
     except AssertionError:
         if must_pass:
             pytest.fail("shouldn't have raised")
-        s = str(sys.exc_info()[1])
+        s = six.text_type(sys.exc_info()[1])
         if not s.startswith("assert"):
             return "AssertionError: " + s
         return s
@@ -607,6 +608,21 @@ class TestAssertionRewrite(object):
             assert 0 == f.a
 
         assert r"where 1 = \n{ \n~ \n}.a" in util._format_lines([getmsg(f)])[0]
+
+    def test_custom_repr_non_ascii(self):
+        def f():
+            class A(object):
+                name = u"ä"
+
+                def __repr__(self):
+                    return self.name.encode("UTF-8")  # only legal in python2
+
+            a = A()
+            assert not a.name
+
+        msg = getmsg(f)
+        assert "UnicodeDecodeError" not in msg
+        assert "UnicodeEncodeError" not in msg
 
 
 class TestRewriteOnImport(object):

